File: bindings/ruby/lib/sone/engine.rb
# frozen_string_literal: true

require "json"

require_relative "errors"
require_relative "native"

module Sone
  # Owns the font registry and the decoded-image cache.
  #
  # Skia's font collection is shared inside an engine, so one engine renders one
  # document at a time and every call here takes the lock. Give each thread its
  # own engine for real parallelism rather than sharing one.
  class Engine
    STATUSES = {
      Native::INVALID_ARGUMENT => ArgumentError,
      Native::IR_ERROR => IrError,
      Native::ASSET_ERROR => AssetError,
      Native::RENDER_ERROR => RenderError
    }.freeze

    # @param base_dir [String] the directory relative asset paths resolve against
    def initialize(base_dir = Dir.pwd)
      @pointer = Native.sone_engine_new(base_dir)
      raise Error, "could not create a sone engine" if @pointer.null?

      @lock = Mutex.new
      @closed = false
      ObjectSpace.define_finalizer(self, self.class.reaper(@pointer))
    end

    # The process-wide engine, used when no explicit one is passed.
    def self.default
      @default ||= new
    end

    def self.version
      Native.sone_version.read_string
    end

    # Frees the handle without capturing the engine, which would keep it alive.
    def self.reaper(pointer)
      proc { Native.sone_engine_free(pointer) }
    end

    def close
      @lock.synchronize do
        next if @closed

        ObjectSpace.undefine_finalizer(self)
        Native.sone_engine_free(@pointer)
        @closed = true
      end
      nil
    end

    def closed?
      @closed
    end

    # ── fonts and assets ────────────────────────────────────────────────────

    # Register a font family from raw TTF/OTF bytes.
    def register_font(name, data)
      with_bytes(data) do |pointer, length|
        call { Native.sone_register_font(handle, name.to_s, pointer, length) }
      end
    end

    # Register a font family from a file.
    def register_font_file(name, path)
      call { Native.sone_register_font_file(handle, name.to_s, path.to_s) }
    end

    # Make bytes available to documents as `asset:<name>`.
    def register_image(name, data)
      with_bytes(data) do |pointer, length|
        call { Native.sone_register_image(handle, name.to_s, pointer, length) }
      end
    end

    def font?(name)
      @lock.synchronize { Native.sone_has_font(handle, name.to_s) }
    end

    def font_families
      JSON.parse(buffer { |out| Native.sone_font_families(handle, out) })
    end

    def reset_fonts
      @lock.synchronize { Native.sone_reset_fonts(handle) }
      nil
    end

    # ── rendering ───────────────────────────────────────────────────────────

    # Render an IR document to bytes.
    def render(document, format: :png, density: nil, quality: 1.0, strict: false)
      options = options_for(format, density, quality, strict)
      buffer { |out| Native.sone_render_json(handle, document, options.pointer, out) }
    end

    # One raster image per page. Requires `pageHeight` in the document config.
    def render_pages(document, format: :png, density: nil, quality: 1.0, strict: false)
      options = options_for(format, density, quality, strict)
      list = Native::BufferList.new
      @lock.synchronize do
        begin
          check(Native.sone_render_pages(handle, document, options.pointer, list))
          list.pages
        ensure
          Native.sone_buffer_list_free(list)
        end
      end
    end

    # The computed layout tree, as a JSON string.
    def dump_layout(document)
      buffer { |out| Native.sone_dump_layout(handle, document, out) }.force_encoding(Encoding::UTF_8)
    end

    # Dataset-style metadata, as a JSON string.
    def dump_metadata(document, granularity = :node)
      buffer { |out| Native.sone_dump_metadata(handle, document, granularity.to_s, out) }
        .force_encoding(Encoding::UTF_8)
    end

    def inspect
      "#<Sone::Engine#{closed? ? " closed" : ""}>"
    end

    private

    def handle
      raise Error, "this engine has been closed" if @closed

      @pointer
    end

    def options_for(format, density, quality, strict)
      code = Native::FORMATS[format.to_s.downcase.to_sym]
      raise ArgumentError, "unknown output format #{format.inspect}" unless code

      options = Native::RenderOptions.new
      options[:format] = code
      # Zero tells the engine to fall back to the document's own config.
      options[:density] = density ? density.to_f : 0.0
      options[:quality] = quality.to_f
      options[:strict] = strict ? 1 : 0
      options
    end

    # Runs a call that fills a buffer, and always releases it.
    def buffer
      out = Native::Buffer.new
      @lock.synchronize do
        begin
          check(yield(out))
          out.bytes
        ensure
          Native.sone_buffer_free(out)
        end
      end
    end

    def call
      @lock.synchronize { check(yield) }
      nil
    end

    def with_bytes(data)
      bytes = data.to_s.b
      FFI::MemoryPointer.new(:uint8, bytes.bytesize) do |pointer|
        pointer.put_bytes(0, bytes)
        return yield(pointer, bytes.bytesize)
      end
    end

    def check(status)
      return status if status == Native::OK

      pointer = Native.sone_engine_last_error(@pointer)
      message = pointer.null? ? "sone failed with status #{status}" : pointer.read_string
      raise (STATUSES[status] || Error), message.force_encoding(Encoding::UTF_8)
    end
  end

  # Font registration on the process-wide engine, for scripts that do not want
  # to own one. Skia carries no system fonts, so at least one family must be
  # registered before any text renders.
  module Font
    module_function

    def load(name, source)
      engine = Engine.default
      if source.respond_to?(:to_path) || (source.is_a?(String) && File.exist?(source))
        engine.register_font_file(name, source.to_s)
      else
        engine.register_font(name, source)
      end
    end

    def has?(name)
      Engine.default.font?(name)
    end

    def families
      Engine.default.font_families
    end

    def reset
      Engine.default.reset_fonts
    end
  end
end

File: bindings/csharp/src/Sone/Engine.cs
using System.Runtime.InteropServices;
using System.Text;
using System.Text.Json;

namespace Sone;

/// <summary>
/// Owns the font registry and the decoded-image cache.
/// </summary>
/// <remarks>
/// Skia's font collection is shared inside an engine, so one engine renders one
/// document at a time — every call here takes a lock. For real parallelism give
/// each thread its own <see cref="Engine"/> rather than sharing one.
/// </remarks>
public sealed class Engine : IDisposable
{
    private static readonly Lazy<Engine> Shared = new(() => new Engine(Environment.CurrentDirectory));

    private readonly object _gate = new();
    private nint _handle;

    /// <param name="baseDir">
    /// The directory relative asset paths resolve against. Null means the
    /// process working directory.
    /// </param>
    public Engine(string? baseDir = null)
    {
        _handle = Native.EngineNew(baseDir);
        if (_handle == nint.Zero)
        {
            throw new SoneException("could not create a sone engine");
        }
    }

    ~Engine() => Free();

    /// <summary>The process-wide engine used when no explicit one is passed.</summary>
    public static Engine Default => Shared.Value;

    /// <summary>The native library version.</summary>
    public static string Version => Marshal.PtrToStringUTF8(Native.Version()) ?? "unknown";

    // ── fonts and assets ─────────────────────────────────────────────────────

    /// <summary>Register a font family from raw TTF/OTF bytes.</summary>
    public unsafe void RegisterFont(string name, ReadOnlySpan<byte> data)
    {
        ArgumentNullException.ThrowIfNull(name);
        lock (_gate)
        {
            fixed (byte* pointer = data)
            {
                Check(Native.RegisterFont(Live, name, pointer, (nuint)data.Length));
            }
        }
    }

    /// <summary>Register a font family from a file.</summary>
    public void RegisterFontFile(string name, string path)
    {
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(path);
        lock (_gate)
        {
            Check(Native.RegisterFontFile(Live, name, path));
        }
    }

    /// <summary>Make bytes available to documents as <c>asset:name</c>.</summary>
    public unsafe void RegisterImage(string name, ReadOnlySpan<byte> data)
    {
        ArgumentNullException.ThrowIfNull(name);
        lock (_gate)
        {
            fixed (byte* pointer = data)
            {
                Check(Native.RegisterImage(Live, name, pointer, (nuint)data.Length));
            }
        }
    }

    /// <summary>Whether a family has been registered.</summary>
    public bool HasFont(string name)
    {
        ArgumentNullException.ThrowIfNull(name);
        lock (_gate)
        {
            return Native.HasFont(Live, name);
        }
    }

    /// <summary>Every registered family name.</summary>
    public IReadOnlyList<string> FontFamilies()
    {
        byte[] json;
        lock (_gate)
        {
            SoneBuffer buffer = default;
            var status = Native.FontFamilies(Live, out buffer);
            json = Collect(status, ref buffer);
        }
        return JsonSerializer.Deserialize(json, SoneJson.Default.StringArray) ?? [];
    }

    /// <summary>Drop every registered font.</summary>
    public void ResetFonts()
    {
        lock (_gate)
        {
            Native.ResetFonts(Live);
        }
    }

    // ── rendering ────────────────────────────────────────────────────────────

    /// <summary>Render an IR document to bytes.</summary>
    public byte[] Render(
        string document,
        OutputFormat format = OutputFormat.Png,
        double? density = null,
        double quality = 1.0,
        bool strict = false) =>
        Render(Utf8(document), format, density, quality, strict);

    /// <summary>One raster image per page. Requires <c>pageHeight</c> in the document config.</summary>
    public IReadOnlyList<byte[]> RenderPages(
        string document,
        OutputFormat format = OutputFormat.Png,
        double? density = null,
        double quality = 1.0,
        bool strict = false) =>
        RenderPages(Utf8(document), format, density, quality, strict);

    /// <summary>The computed layout tree, as JSON.</summary>
    public string DumpLayout(string document) => DumpLayout(Utf8(document));

    /// <summary>Dataset-style metadata, as JSON.</summary>
    public string DumpMetadata(string document, string granularity = "node") =>
        DumpMetadata(Utf8(document), granularity);

    // The overloads Rendering uses: the document is already NUL-terminated
    // UTF-8, so nothing re-encodes on the way to the engine.

    internal unsafe byte[] Render(byte[] document, OutputFormat format, double? density, double quality, bool strict)
    {
        var options = OptionsFor(format, density, quality, strict);
        lock (_gate)
        {
            fixed (byte* json = document)
            {
                SoneBuffer buffer = default;
                var status = Native.RenderJson(Live, json, in options, out buffer);
                return Collect(status, ref buffer);
            }
        }
    }

    internal unsafe IReadOnlyList<byte[]> RenderPages(byte[] document, OutputFormat format, double? density, double quality, bool strict)
    {
        var options = OptionsFor(format, density, quality, strict);
        lock (_gate)
        {
            fixed (byte* json = document)
            {
                SoneBufferList list = default;
                var status = Native.RenderPages(Live, json, in options, out list);
                try
                {
                    Check(status);
                    var items = (SoneBuffer*)list.Items;
                    var pages = new byte[(int)list.Len][];
                    for (var i = 0; i < pages.Length; i++)
                    {
                        pages[i] = Copy(items[i]);
                    }
                    return pages;
                }
                finally
                {
                    Native.BufferListFree(ref list);
                }
            }
        }
    }

    internal unsafe string DumpLayout(byte[] document)
    {
        lock (_gate)
        {
            fixed (byte* json = document)
            {
                SoneBuffer buffer = default;
                var status = Native.DumpLayout(Live, json, out buffer);
                return Encoding.UTF8.GetString(Collect(status, ref buffer));
            }
        }
    }

    internal unsafe string DumpMetadata(byte[] document, string granularity)
    {
        lock (_gate)
        {
            fixed (byte* json = document)
            {
                SoneBuffer buffer = default;
                var status = Native.DumpMetadata(Live, json, granularity, out buffer);
                return Encoding.UTF8.GetString(Collect(status, ref buffer));
            }
        }
    }

    // ── lifetime ─────────────────────────────────────────────────────────────

    public void Dispose()
    {
        Free();
        GC.SuppressFinalize(this);
    }

    private void Free()
    {
        lock (_gate)
        {
            if (_handle != nint.Zero)
            {
                Native.EngineFree(_handle);
                _handle = nint.Zero;
            }
        }
    }

    // ── internals ────────────────────────────────────────────────────────────

    /// <summary>The handle, or a clear error rather than a segfault.</summary>
    private nint Live
    {
        get
        {
            ObjectDisposedException.ThrowIf(_handle == nint.Zero, this);
            return _handle;
        }
    }

    private static SoneRenderOptions OptionsFor(OutputFormat format, double? density, double quality, bool strict) => new()
    {
        Format = format,
        // Zero tells the engine to fall back to the document's own config.
        Density = (float)(density ?? 0.0),
        Quality = (float)quality,
        Strict = strict ? 1 : 0,
    };

    /// <summary>A NUL-terminated UTF-8 copy, which is what the C ABI takes.</summary>
    internal static byte[] Utf8(string document)
    {
        var bytes = new byte[Encoding.UTF8.GetByteCount(document) + 1];
        Encoding.UTF8.GetBytes(document, bytes);
        return bytes;
    }

    private byte[] Collect(SoneStatus status, ref SoneBuffer buffer)
    {
        try
        {
            Check(status);
            return Copy(buffer);
        }
        finally
        {
            Native.BufferFree(ref buffer);
        }
    }

    private static byte[] Copy(SoneBuffer buffer)
    {
        if (buffer.Data == nint.Zero || buffer.Len == 0)
        {
            return [];
        }
        var bytes = new byte[(int)buffer.Len];
        Marshal.Copy(buffer.Data, bytes, 0, bytes.Length);
        return bytes;
    }

    private void Check(SoneStatus status)
    {
        if (status == SoneStatus.Ok)
        {
            return;
        }
        var message = Marshal.PtrToStringUTF8(Native.EngineLastError(_handle))
                      ?? $"sone failed with {status}";
        throw status switch
        {
            SoneStatus.InvalidArgument => new ArgumentException(message),
            SoneStatus.IrError => new IrException(message),
            SoneStatus.AssetError => new AssetException(message),
            _ => (Exception)new RenderException(message),
        };
    }
}
